Fix crash in git_clone on extremely large repos

jeffhostetler opened the pull request:
A crash was observed on Windows when cloning a very large repo. In this instance, the received packfile was nearly 6Gb.

The problem was in sloppy casting of the "size" field from src/indexer.c:append_to_pack() to src/win32/posix_w32.c:p_ftruncate() to _chsize_s().
Where the file position was being truncated and then sign-extended.

Also fixed the return value from _chsize_s() which differs from _chsize().

jeffhostetler commented:
Looks like this breaks ming-win32 which (still) doesn't have _chsize_s().  Need to investigate this a bit.

jeffhostetler commented:
I re-worked the change in p_ftruncate() for Windows and restored the original code calling _chsize() which appears to only be needed by ming-w32.  

However, I did add argument checking. 

I'm wondering if we want to have simple clar test that just calls p_ftruncate(fd, 2Gb+1 or maybe 4Gb+1) to confirm this in isolation.

ethomson commented:
That'd be great; it would be nice to have a test that tests the problem and not something that depends on a big repo.  We have some tests that write big files (defaulting to off), looking at the `GITTEST_INVASIVE_FS_SIZE` environment variable.  (One of the filter tests.)

> On Feb 28, 2015, at 9:57 AM, Jeff Hostetler <notifications@github.com> wrote:
> 
> I re-worked the change in p_ftruncate() for Windows and restored the original code calling _chsize() which appears to only be needed by ming-w32.
> 
> However, I did add argument checking.
> 
> I'm wondering if we want to have simple clar test that just calls p_ftruncate(fd, 2Gb+1 or maybe 4Gb+1) to confirm this in isolation.
> 
> —
> Reply to this email directly or view it on GitHub.
>

carlosmn commented on src/win32/posix_w32.c:
```diff
@@ -42,12 +42,26 @@
+	 *      This comment is probably __MINGW32__ specific.
+	 */
+	if (size > 0x7fffffff)
+		return -1;
```
The caller would inspect `errno` if this function fails, no? We should set it to something useful

jeffhostetler replied:
good catch. i'll set errno to something sane like the actual _chsize() would.

carlosmn commented:
Yeah, let's go for `GITTEST_INVASIVE_FS_SIZE` and create one or two large objects from a random source which we can then clone locally

jeffhostetler commented:
I've added a simple "open(); ftruncate(); lstat()" unit test to verify that the ftruncate() works as expected for big files.

jeffhostetler commented:
I'm adding another test to "clone empty_test_repo", stage and commit a huge file, and locally clone it
but ran into https://github.com/libgit2/libgit2/issues/2936

I think I'll commit my simple unit test for ftruncate and let us close this problem and not look for a solution to the blog problem in this PR.

carlosmn commented on tests/core/ftruncate.c:
```diff
@@ -0,0 +1,50 @@
+	struct stat st;
+	int error;
+
+	assert(sizeof(i64len) == 8); /* yes, i'm that paranoid. */
```
We do guarantee this through the headers (which is why we don't use `off_t`).

carlosmn commented on src/win32/posix_w32.c:
```diff
@@ -42,12 +42,30 @@
+	/* TODO Find a replacement for _chsize() that handles big files.
+	 *      This comment is probably __MINGW32__ specific.
+	 */
+	if (size > 0x7fffffff) {
```
We should have `INT32_MAX` defined in any OS, it'd be more descriptive to use that name.

jeffhostetler commented:
OK. I fixed the constants and squashed everything.  With that it now just looks like a fix to p_ftruncate() with nothing to indicate it was discovered via git_clone().  So I'd like to go ahead and mark this finished.  If/when I get time, I'll do another PR containing a unit test to build upon this and try to clone a huge repo.

ethomson commented on src/win32/posix_w32.c:
```diff
@@ -42,12 +42,30 @@
+		return -1;
+	}
+
 #if defined(_MSC_VER) && _MSC_VER >= 1500
```
I think that this (historic) test is incorrect.  `_chsize_s` was available way back in XP.  Should this just be `#ifdef __MINGW32__` ?

carlosmn commented:
Agreed, let's just go with the truncate fix. We can have some PR creating three or four 2GB objects which we can then clone, but allowing the creation of an object larger than `SIZE_MAX` is going to require quite some changes.
